src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java

```diff
@@ -0,0 +1,429 @@
+                assertEquals(expectedNullable, rsMetaData.isNullable(i + 1));
+            }
+        }
+        if (test.expected_is_read_only != null) {
+            assertEquals(test.expected_is_read_only.size(), columnCount);
+            for (int i = 0; i < columnCount; i++) {
+                assertEquals(test.expected_is_read_only.get(i), rsMetaData.isReadOnly(i + 1));
+            }
+        }
+        if (test.expected_is_searchable != null) {
+            assertEquals(test.expected_is_searchable.size(), columnCount);
+            for (int i = 0; i < columnCount; i++) {
+                assertEquals(test.expected_is_searchable.get(i), rsMetaData.isSearchable(i + 1));
+            }
+        }
+        if (test.expected_is_signed != null) {
+            assertEquals(test.expected_is_signed.size(), columnCount);
+            for (int i = 0; i < columnCount; i++) {
+                assertEquals(test.expected_is_signed.get(i), rsMetaData.isSigned(i + 1));
+            }
+        }
+        if (test.expected_is_writable != null) {
+            assertEquals(test.expected_is_writable.size(), columnCount);
+            for (int i = 0; i < columnCount; i++) {
+                assertEquals(test.expected_is_writable.get(i), rsMetaData.isWritable(i + 1));
+            }
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private static void validateResultsOrdered(TestEntry testEntry, ResultSet rs)
+            throws SQLException {
+        Integer actualRowCounter = null;
+        if (testEntry.expected_result != null) {
+            while (rs.next()) {
+                if (actualRowCounter == null) {
+                    actualRowCounter = 0;
+                }
+                assertTrue(
+                        compareRow(
+                                (List<Object>) testEntry.expected_result.get(actualRowCounter),
+                                rs));
+                actualRowCounter++;
+            }
+        }
+        if (testEntry.row_count != null) {
+            validateRowCount(testEntry.row_count, testEntry, actualRowCounter, rs);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private static void validateResultsUnordered(TestEntry testEntry, ResultSet rs)
+            throws SQLException {
+        Integer actualRowCounter = null;
+        if (testEntry.expected_result != null) {
+            while (rs.next()) {
+                boolean found = false;
+                if (actualRowCounter == null) {
+                    actualRowCounter = 0;
+                }
+                actualRowCounter++;
+                for (Object expectedRow : testEntry.expected_result) {
+                    if (compareRow((List<Object>) expectedRow, rs)) {
+                        found = true;
+                        break;
+                    }
+                }
+                assertTrue(found);
+            }
+        }
+        if (testEntry.row_count != null) {
```

Comment:
The `validateRowCount` method uses the value `actualRowCounter` which can be set during the row validation step.  
If not set, then the validateRowCount iterates through the rowset and counts the rows.  One way I can think of to move validateRowCount above the row validation is to have copy the results to a list as it iterates through the rows and use that list in the row validation.  I'm not sure if that adds too much value as it adds having to copy the results.  WDYT?

Reply:
oh, no, sorry, I forgot that we don't know the row count and have to go through the whole rs. 
There is one thing however that you might need to address : Technically, if I only want to validate the row count and not the rows themselves, it won't work right now because it will not go through the result set unless testEntry.expected_result != null, correct?

Reply:
No problem.
If `actualRowCounter` is still null then the `validateRowCount()` method will go ahead and count the rows of the resultset itself [here](https://github.com/mongodb/mongo-jdbc-driver/pull/62/files#diff-e017435f177b34f6c79fba493b695e25515551ac3ac287cc8554c7257bababbbR365).
Now that I'm looking at `actualRowCounter` I see I should move it's initial null check to before the `while (rs.next()) {` loop, I'll do that now.